pkg/diagnostics/http_monitoring.go

```diff
@@ -307,3 +346,94 @@ func (h *httpMetrics) HTTPMiddleware(next http.Handler) http.Handler {
 		h.ServerRequestCompleted(r.Context(), method, path, status, reqContentSize, respSize, elapsed)
 	})
 }
+
+func (h *httpMetrics) initPathMatching(config *config.PathMatching) pathMatchingConfig {
+	if config == nil {
+		return pathMatchingConfig{
+			enabled: false,
+		}
+	}
+
+	pm := pathMatchingConfig{
+		enabled: len(config.EgressPaths)+len(config.IngressPaths) > 0,
+	}
+
+	if !pm.enabled {
+		return pm
+	}
+
+	pm.virtualIngressMux = http.NewServeMux()
+	pm.virtualIngressMux.Handle("/", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if !h.legacy {
+			rw, ok := w.(*diagUtils.PathMatchingRW)
+			if !ok {
+				log.Errorf("Failed to cast to PathMatchingRW")
+				return
+			}
+			rw.MatchedPath = diagUtils.UnmatchedPathPlaceholder
+		}
+	}))
+	for _, pattern := range config.IngressPaths {
+		pm.virtualIngressMux.Handle(pattern, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			rw, ok := w.(*diagUtils.PathMatchingRW)
+			if !ok {
+				log.Errorf("Failed to cast to PathMatchingRW")
+				return
+			}
+			rw.MatchedPath = pattern
+		}))
+	}
```

Comment:
Rather than passing around strings and doing any/interface conversions- can we put the `stats.RecordWithTags` inside these request handlers?

Reply:
I don't think we can avoid the interface conversions here since we always need to cast the appropriate response writer 🙁 

Moving the metric producers inside the handlers would require us to have one rw per metric type. Personally, I'd prefer to keep things as they are now and not introduce that change in this PR if you are OK.